# 1. Introducción teórica

## Aprendizaje automático y análisis de datos

El aprendizaje automático (*Machine Learning*) es una de las disciplinas más relevantes dentro de la inteligencia artificial, ya que permite desarrollar modelos capaces de identificar patrones, realizar predicciones y descubrir relaciones significativas a partir de grandes volúmenes de información. Dependiendo de la naturaleza de los datos disponibles y del objetivo que se persiga, los algoritmos de aprendizaje pueden clasificarse principalmente en **aprendizaje supervisado** y **aprendizaje no supervisado**.

---

## Aprendizaje supervisado

El aprendizaje supervisado se basa en el uso de conjuntos de datos previamente etiquetados durante el proceso de entrenamiento. Esto significa que cada registro contiene una variable objetivo conocida, lo que permite al algoritmo aprender la relación existente entre las variables de entrada y la respuesta esperada.

Una vez finalizada la fase de entrenamiento, el modelo es capaz de generalizar el conocimiento adquirido y realizar predicciones sobre datos nuevos que no han sido observados previamente.

Este enfoque se utiliza principalmente en dos tipos de problemas:

### Clasificación

Consiste en determinar la categoría o clase a la que pertenece una observación determinada.

**Ejemplos:**

* Identificar si una conexión de red corresponde a tráfico normal o malicioso.
* Clasificar correos electrónicos como spam o legítimos.
* Determinar si una imagen contiene un perro o un gato.

### Regresión

Busca estimar valores numéricos continuos a partir de un conjunto de variables predictoras.

**Ejemplos:**

* Predicción del precio de una vivienda.
* Estimación del consumo energético.
* Pronóstico de ventas futuras.

### Algoritmos supervisados más representativos

Entre los algoritmos de aprendizaje supervisado más utilizados en la literatura destacan:

* Regresión Lineal (*Linear Regression*)
* Regresión Logística (*Logistic Regression*)
* Árboles de Decisión (*Decision Trees*)
* Bosques Aleatorios (*Random Forests*)
* Máquinas de Vectores de Soporte (*Support Vector Machines - SVM*)
* K-Nearest Neighbors (*KNN*)
* Redes Neuronales Artificiales (*Artificial Neural Networks*)

---

## Aprendizaje no supervisado

A diferencia del aprendizaje supervisado, el aprendizaje no supervisado trabaja con datos que no poseen una variable objetivo previamente definida. En este escenario, el algoritmo no recibe una respuesta correcta durante el entrenamiento, sino que explora la información disponible para identificar estructuras internas, relaciones ocultas o agrupamientos naturales entre los datos.

Este tipo de aprendizaje resulta especialmente útil cuando se busca comprender mejor la estructura de un conjunto de datos o descubrir patrones que no son evidentes a simple vista.

### Aplicaciones comunes

Algunos ejemplos de aplicación incluyen:

* Segmentación de clientes según sus hábitos de consumo.
* Sistemas de recomendación de contenido como los utilizados por Netflix o Spotify.
* Agrupamiento automático de documentos y noticias similares.
* Detección de anomalías o comportamientos atípicos.

### Algoritmos no supervisados más representativos

Entre los algoritmos más utilizados se encuentran:

* K-Means
* Agrupamiento Jerárquico (*Hierarchical Clustering*)
* DBSCAN (*Density-Based Spatial Clustering of Applications with Noise*)
* Modelos de Mezcla Gaussiana (*Gaussian Mixture Models - GMM*)
* Análisis de Componentes Principales (*Principal Component Analysis - PCA*)

---

## Algoritmos de aprendizaje automático disponibles en PySpark

Para el procesamiento de grandes volúmenes de información, Apache Spark proporciona la biblioteca **MLlib**, accesible desde PySpark mediante el módulo `pyspark.ml`. Esta biblioteca incorpora implementaciones distribuidas de numerosos algoritmos de aprendizaje automático, permitiendo trabajar de manera eficiente con conjuntos de datos de gran tamaño.

### Algoritmos supervisados disponibles en PySpark

* `DecisionTreeClassifier`
* `DecisionTreeRegressor`
* `RandomForestClassifier`
* `RandomForestRegressor`
* `GBTClassifier`
* `GBTRegressor`
* `MultilayerPerceptronClassifier`
* `LogisticRegression`
* `LinearRegression`
* `LinearSVC`
* `NaiveBayes`

### Algoritmos no supervisados disponibles en PySpark

* `KMeans`
* `GaussianMixture`
* `BisectingKMeans`
* `PowerIterationClustering`
* `LatentDirichletAllocation (LDA)`
* `PCA`

---

## Aplicación al proyecto

Para el desarrollo de esta actividad se utilizará una muestra obtenida a partir del conjunto de datos **Containers_Dataset**, perteneciente al repositorio **Integrated Cloud Security Datasets**, el cual contiene información relacionada con flujos de tráfico de red generados en entornos Kubernetes y escenarios de seguridad en la nube.

A partir de esta información se llevarán a cabo dos experimentos independientes:

1. **Aprendizaje supervisado**, utilizando la variable **Label** como variable objetivo para realizar tareas de clasificación.
2. **Aprendizaje no supervisado**, empleando variables representativas del tráfico de red para identificar agrupamientos naturales dentro de los datos.

Para garantizar resultados confiables, previamente se realizarán las etapas de selección, preparación y partición de los datos, asegurando que la muestra utilizada sea representativa y adecuada para el entrenamiento y evaluación de los modelos implementados mediante PySpark.

# 2. Selección de los Datos y Construcción de la Muestra M


## Descripción de la población de estudio

La población de datos utilizada en este proyecto corresponde al conjunto Containers_Dataset, perteneciente al repositorio Integrated Cloud Security Datasets. Este conjunto contiene información de flujos de red capturados en entornos Kubernetes que ejecutan aplicaciones basadas en contenedores y microservicios. Los registros incluyen tanto tráfico legítimo como distintos tipos de actividades asociadas a ataques de seguridad informática.

Dado que el conjunto completo contiene un volumen considerable de observaciones, utilizar la totalidad de los registros durante las etapas de experimentación incrementaría significativamente los tiempos de procesamiento. Por esta razón, se decidió construir una muestra representativa de tamaño reducido que permitiera realizar las pruebas de aprendizaje automático de manera eficiente sin perder representatividad estadística.

# Variables de caracterización utilizadas

Tomando como referencia la actividad desarrollada en el Módulo 3, se identificaron dos variables de caracterización principales:

traffic_type: clasifica el tráfico como benigno o malicioso.
protocol_type: identifica el protocolo de red utilizado por cada flujo.

Estas variables permiten dividir la población original en particiones homogéneas, garantizando que los diferentes comportamientos presentes en la población sean representados dentro de la muestra final.

# Estrategia de muestreo

Para construir la muestra M, se aplicó una estrategia de muestreo estratificado, donde cada estrato corresponde a una combinación de los valores de las variables de caracterización identificadas previamente.

El objetivo de esta técnica fue evitar que las clases con mayor cantidad de registros dominaran la muestra final, permitiendo conservar una representación equilibrada de los distintos tipos de tráfico presentes en la base de datos.

Posteriormente, se recuperó una cantidad limitada de registros de cada estrato y se realizó la unión de todos ellos para formar la muestra final M, la cual fue utilizada en las etapas posteriores de preparación de datos, entrenamiento y evaluación de modelos.

In [1]:
%pip install pyspark
%pip install kaggle
%pip install --upgrade kaggle kagglehub kagglesdk
import os
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, input_file_name, regexp_extract, lit

Note: you may need to restart the kernel to use updated packages.
  Using cached kaggle-2.2.0-py3-none-any.whl.metadata (16 kB)
  Using cached jupytext-1.19.3-py3-none-any.whl.metadata (15 kB)
  Using cached kagglesdk-0.1.27-py3-none-any.whl.metadata (13 kB)
Using cached kaggle-2.2.0-py3-none-any.whl (128 kB)
Using cached kagglesdk-0.1.27-py3-none-any.whl (229 kB)
Using cached jupytext-1.19.3-py3-none-any.whl (170 kB)

   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ---------------------------------------- 0/3 [kagglesdk]
   ----------------------------------

In [4]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, when, trim

spark = SparkSession.builder \
    .appName("ContainersDataset") \
    .getOrCreate()

print("Spark UI:", spark.sparkContext.uiWebUrl)

df = spark.read.csv(
    r"C:\Users\masalin2\Downloads\Containers_Dataset.csv",
    header=True,
    inferSchema=False
)

df.select("Label", "Protocol").show(10, truncate=False)

df_particiones = df.withColumn(
    "traffic_type",
    when(trim(col("Label").cast("string")) == "0", "Benign")
    .otherwise("Malicious")
)

df_particiones = df_particiones.withColumn(
    "protocol_num",
    col("Protocol").cast("int")
)

df_particiones = df_particiones.withColumn(
    "protocol_type",
    when(col("protocol_num") == 6, "TCP")
    .when(col("protocol_num") == 17, "UDP")
    .otherwise("Other")
)

df_particiones.groupBy(
    "traffic_type",
    "protocol_type"
).count().show()

muestra_M = (
    df_particiones
    .sampleBy(
        "traffic_type",
        fractions={
            "Benign": 0.05,
            "Malicious": 0.05
        },
        seed=42
    )
)

print("Registros en M:", muestra_M.count())

Spark UI: http://host.docker.internal:4040
+-----+--------+
|Label|Protocol|
+-----+--------+
|0    |0       |
|0    |6       |
|0    |0       |
|0    |0       |
|0    |6       |
|0    |0       |
|0    |6       |
|0    |6       |
|0    |6       |
|0    |6       |
+-----+--------+
only showing top 10 rows
+------------+-------------+-------+
|traffic_type|protocol_type|  count|
+------------+-------------+-------+
|      Benign|        Other|   5235|
|      Benign|          TCP|2905606|
|      Benign|          UDP|  42450|
|   Malicious|        Other|   2524|
|   Malicious|          UDP|  21221|
|   Malicious|          TCP| 254439|
+------------+-------------+-------+

Registros en M: 161672


### Interpretación de la selección de datos

A partir del conjunto global D se generaron particiones utilizando dos variables de caracterización: `traffic_type` y `protocol_type`. La variable `traffic_type` fue derivada de la columna `Label`, donde el valor 0 representa tráfico benigno y los demás valores se consideran tráfico malicioso. Por su parte, `protocol_type` fue derivada de la columna `Protocol`, clasificando los valores 6 como TCP, 17 como UDP y el resto como Other.

Los resultados muestran que la población contiene registros benignos y maliciosos distribuidos en distintos protocolos de red. Posteriormente, se aplicó un muestreo estratificado para construir la muestra M, reduciendo el volumen de datos original a 161,672 registros. Esta muestra conserva las particiones principales identificadas en la población, permitiendo trabajar con un conjunto de datos más manejable sin perder representatividad para las etapas posteriores de preparación, entrenamiento y evaluación de modelos.

## Preparación de los datos

In [5]:
print("Número de columnas:", len(muestra_M.columns))
print("Número de registros:", muestra_M.count())

Número de columnas: 90
Número de registros: 161672


In [6]:
muestra_M.printSchema()

root
 |-- Flow ID: string (nullable = true)
 |-- Src IP: string (nullable = true)
 |-- Src Port: string (nullable = true)
 |-- Dst IP: string (nullable = true)
 |-- Dst Port: string (nullable = true)
 |-- Protocol: string (nullable = true)
 |-- Timestamp: string (nullable = true)
 |-- Flow Duration: string (nullable = true)
 |-- Total Fwd Packet: string (nullable = true)
 |-- Total Bwd packets: string (nullable = true)
 |-- Total Length of Fwd Packet: string (nullable = true)
 |-- Total Length of Bwd Packet: string (nullable = true)
 |-- Fwd Packet Length Max: string (nullable = true)
 |-- Fwd Packet Length Min: string (nullable = true)
 |-- Fwd Packet Length Mean: string (nullable = true)
 |-- Fwd Packet Length Std: string (nullable = true)
 |-- Bwd Packet Length Max: string (nullable = true)
 |-- Bwd Packet Length Min: string (nullable = true)
 |-- Bwd Packet Length Mean: string (nullable = true)
 |-- Bwd Packet Length Std: string (nullable = true)
 |-- Flow Bytes/s: string (nullable

In [7]:
from pyspark.sql.functions import col, sum as spark_sum

muestra_M.select([
    spark_sum(col(c).isNull().cast("int")).alias(c)
    for c in muestra_M.columns
]).show(vertical=True, truncate=False)

-RECORD 0-------------------------
 Flow ID                    | 0   
 Src IP                     | 0   
 Src Port                   | 0   
 Dst IP                     | 0   
 Dst Port                   | 0   
 Protocol                   | 0   
 Timestamp                  | 0   
 Flow Duration              | 0   
 Total Fwd Packet           | 0   
 Total Bwd packets          | 0   
 Total Length of Fwd Packet | 0   
 Total Length of Bwd Packet | 0   
 Fwd Packet Length Max      | 0   
 Fwd Packet Length Min      | 0   
 Fwd Packet Length Mean     | 0   
 Fwd Packet Length Std      | 0   
 Bwd Packet Length Max      | 0   
 Bwd Packet Length Min      | 0   
 Bwd Packet Length Mean     | 0   
 Bwd Packet Length Std      | 0   
 Flow Bytes/s               | 0   
 Flow Packets/s             | 0   
 Flow IAT Mean              | 0   
 Flow IAT Std               | 0   
 Flow IAT Max               | 0   
 Flow IAT Min               | 0   
 Fwd IAT Total              | 0   
 Fwd IAT Mean       

In [8]:
from pyspark.sql.functions import col

muestra_prep = (
    muestra_M
    .withColumn("Flow Duration", col("Flow Duration").cast("double"))
    .withColumn("Total Fwd Packet", col("Total Fwd Packet").cast("double"))
    .withColumn("Total Bwd packets", col("Total Bwd packets").cast("double"))
    .withColumn("Flow Bytes/s", col("Flow Bytes/s").cast("double"))
    .withColumn("Flow Packets/s", col("Flow Packets/s").cast("double"))
    .withColumn("Packet Length Mean", col("Packet Length Mean").cast("double"))
    .withColumn("Packet Length Std", col("Packet Length Std").cast("double"))
)

muestra_prep.select(
    "Flow Duration",
    "Total Fwd Packet",
    "Flow Bytes/s"
).printSchema()

root
 |-- Flow Duration: double (nullable = true)
 |-- Total Fwd Packet: double (nullable = true)
 |-- Flow Bytes/s: double (nullable = true)



In [9]:
from pyspark.sql.functions import sum as spark_sum

muestra_prep.select([
    spark_sum(col(c).isNull().cast("int")).alias(c)
    for c in [
        "Flow Duration",
        "Total Fwd Packet",
        "Total Bwd packets",
        "Flow Bytes/s",
        "Flow Packets/s",
        "Packet Length Mean",
        "Packet Length Std"
    ]
]).show()

+-------------+----------------+-----------------+------------+--------------+------------------+-----------------+
|Flow Duration|Total Fwd Packet|Total Bwd packets|Flow Bytes/s|Flow Packets/s|Packet Length Mean|Packet Length Std|
+-------------+----------------+-----------------+------------+--------------+------------------+-----------------+
|            0|               0|                0|           0|             0|                 0|                0|
+-------------+----------------+-----------------+------------+--------------+------------------+-----------------+



### Resultados de la preparación de los datos

Durante la etapa de preparación se verificó la calidad de la muestra M, compuesta por 161,672 registros y 90 variables. El análisis realizado mostró que no existen valores nulos en ninguna de las columnas, por lo que no fue necesario aplicar técnicas de imputación o eliminación de registros.

Posteriormente, se revisaron los tipos de datos de las variables y se identificó que la mayoría de las características se encontraban almacenadas como cadenas de texto (string). Debido a que los algoritmos de aprendizaje automático requieren variables numéricas para su procesamiento, se realizó la conversión de las características más relevantes a tipo double.

Finalmente, se consideró la presencia potencial de valores atípicos. Sin embargo, debido a la naturaleza del conjunto de datos, asociado al análisis de tráfico de red y detección de actividades maliciosas, estos valores pueden representar comportamientos legítimos o eventos de seguridad relevantes, por lo que no fueron eliminados durante esta etapa.

Con estas acciones se obtuvo una muestra preprocesada y lista para ser utilizada en la construcción de los conjuntos de entrenamiento y prueba.

## Preparación del conjunto de entrenamiento y prueba

In [10]:
# División de la muestra preprocesada
train_df, test_df = muestra_prep.randomSplit(
    [0.8, 0.2],
    seed=42
)

# Verificación de tamaños
print("Registros entrenamiento:", train_df.count())
print("Registros prueba:", test_df.count())

# Porcentajes obtenidos
total = train_df.count() + test_df.count()

print(
    "Porcentaje entrenamiento:",
    round(train_df.count() / total * 100, 2),
    "%"
)

print(
    "Porcentaje prueba:",
    round(test_df.count() / total * 100, 2),
    "%"
)

Registros entrenamiento: 129580
Registros prueba: 32092
Porcentaje entrenamiento: 80.15 %
Porcentaje prueba: 19.85 %


### Preparación del conjunto de entrenamiento y prueba

Una vez concluida la etapa de preprocesamiento, la muestra M fue dividida en dos subconjuntos independientes: entrenamiento y prueba. Para realizar esta partición se utilizó la función `randomSplit()` de PySpark, empleando una semilla fija (`seed=42`) con el objetivo de garantizar la reproducibilidad de los resultados.

Se adoptó una proporción de 80% para entrenamiento y 20% para prueba. Esta distribución es ampliamente utilizada en problemas de aprendizaje automático debido a que permite disponer de una cantidad suficiente de observaciones para el aprendizaje de los modelos, conservando al mismo tiempo un conjunto independiente para evaluar su capacidad de generalización.

La selección aleatoria de los registros contribuye a minimizar la introducción de sesgos durante el proceso de partición, favoreciendo que ambos subconjuntos mantengan una distribución similar a la observada en la muestra original. El conjunto de entrenamiento será utilizado para construir los modelos de aprendizaje supervisado y no supervisado, mientras que el conjunto de prueba permitirá evaluar el desempeño obtenido sobre datos no utilizados durante el entrenamiento.


## Construccion de modelos

### Aprendizaje Supervisado - Random Forest

### Entrenamiento del modelo Random Forest

In [16]:
from pyspark.sql.functions import col, isnan, isnull, when, count
from pyspark.ml.feature import VectorAssembler, StringIndexer
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

feature_cols = [
    "Flow Duration",
    "Total Fwd Packet",
    "Total Bwd packets",
    "Flow Bytes/s",
    "Flow Packets/s",
    "Packet Length Mean",
    "Packet Length Std"
]

# Eliminar NaN, Null e Infinity
condicion_valida = None

for c in feature_cols:
    condicion_columna = (
        col(c).isNotNull() &
        (~isnan(col(c))) &
        (col(c) != float("inf")) &
        (col(c) != float("-inf"))
    )
    
    if condicion_valida is None:
        condicion_valida = condicion_columna
    else:
        condicion_valida = condicion_valida & condicion_columna

muestra_modelo = muestra_prep.filter(condicion_valida)

print("Registros antes de limpieza:", muestra_prep.count())
print("Registros después de limpieza:", muestra_modelo.count())

# Nueva división entrenamiento / prueba
train_df, test_df = muestra_modelo.randomSplit([0.8, 0.2], seed=42)

print("Registros entrenamiento:", train_df.count())
print("Registros prueba:", test_df.count())

# Vector de características
assembler = VectorAssembler(
    inputCols=feature_cols,
    outputCol="features"
)

train_supervised = assembler.transform(train_df)
test_supervised = assembler.transform(test_df)

# Convertir Label a variable objetivo numérica
label_indexer = StringIndexer(
    inputCol="Label",
    outputCol="label"
)

label_model = label_indexer.fit(train_supervised)

train_supervised = label_model.transform(train_supervised)
test_supervised = label_model.transform(test_supervised)

# Entrenar Random Forest
rf = RandomForestClassifier(
    labelCol="label",
    featuresCol="features",
    numTrees=50,
    seed=42
)

rf_model = rf.fit(train_supervised)

# Predicciones
predictions = rf_model.transform(test_supervised)

# Evaluación
accuracy_evaluator = MulticlassClassificationEvaluator(
    labelCol="label",
    predictionCol="prediction",
    metricName="accuracy"
)

accuracy = accuracy_evaluator.evaluate(predictions)

print("Accuracy:", accuracy)

predictions.select("label", "prediction").show(10)

Registros antes de limpieza: 161672
Registros después de limpieza: 159568
Registros entrenamiento: 127915
Registros prueba: 31653
Accuracy: 0.9588348655735633
+-----+----------+
|label|prediction|
+-----+----------+
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
|  0.0|       0.0|
+-----+----------+
only showing top 10 rows


### Discusión de resultados – Modelo de aprendizaje supervisado

Para la construcción del modelo de aprendizaje supervisado se utilizó el algoritmo **Random Forest Classifier**, empleando como variable objetivo la columna **Label** y utilizando siete características numéricas relacionadas con el comportamiento de los flujos de red: *Flow Duration*, *Total Fwd Packet*, *Total Bwd packets*, *Flow Bytes/s*, *Flow Packets/s*, *Packet Length Mean* y *Packet Length Std*.

Durante la etapa de preparación se identificaron registros que contenían valores especiales (*NaN* e *Infinity*), los cuales no pueden ser procesados por los algoritmos de Machine Learning de PySpark. Por esta razón, se eliminaron dichos registros antes del entrenamiento, reduciendo la muestra de **161,672** a **159,568** observaciones, lo que representa una pérdida mínima de información y permite garantizar la calidad de los datos utilizados.

Posteriormente, la muestra fue dividida en un conjunto de entrenamiento de **127,915 registros** y un conjunto de prueba de **31,653 registros**, manteniendo una proporción aproximada de 80% y 20%, respectivamente.

El modelo entrenado obtuvo una métrica de **Accuracy = 0.9588**, equivalente a un **95.88% de exactitud** sobre el conjunto de prueba. Este resultado indica que el modelo fue capaz de clasificar correctamente la gran mayoría de los registros evaluados, mostrando una alta capacidad para identificar los patrones presentes en los datos de tráfico de red.

Al analizar las primeras predicciones generadas, se observa una coincidencia entre las etiquetas reales y las predicciones del modelo, lo que confirma que el proceso de entrenamiento fue exitoso. En términos generales, los resultados sugieren que las variables seleccionadas contienen información relevante para diferenciar los distintos comportamientos presentes en el conjunto de datos y que el algoritmo Random Forest constituye una alternativa adecuada para tareas de clasificación en escenarios de ciberseguridad y análisis de tráfico de red.


## Aprendizaje No Supervisado - K-Means

### Construcción del conjunto de características

In [17]:
from pyspark.ml.feature import VectorAssembler

assembler_kmeans = VectorAssembler(
    inputCols=feature_cols,
    outputCol="features"
)

train_unsupervised = assembler_kmeans.transform(train_df)

## Entrenamiento del modelo K-Means

### Utilizaremos tres grupos inicialmente.

In [18]:
from pyspark.ml.clustering import KMeans

kmeans = KMeans(
    k=3,
    seed=42,
    featuresCol="features"
)

kmeans_model = kmeans.fit(train_unsupervised)

## Asignación de clusters

In [19]:
clusters = kmeans_model.transform(train_unsupervised)

clusters.select(
    "prediction"
).show(20)

+----------+
|prediction|
+----------+
|         0|
|         1|
|         1|
|         1|
|         1|
|         1|
|         1|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
|         0|
+----------+
only showing top 20 rows


## Valuación del agrupamiento

PySpark permite calcular el índice Silhouette.

In [20]:
from pyspark.ml.evaluation import ClusteringEvaluator

evaluator = ClusteringEvaluator()

silhouette = evaluator.evaluate(clusters)

print("Silhouette Score:", silhouette)

Silhouette Score: 0.9899468013209941


## Centros de los clusters

In [21]:
centers = kmeans_model.clusterCenters()

for i, center in enumerate(centers):
    print(f"Cluster {i}")
    print(center)
    print()

Cluster 0
[1.48759722e+06 2.19235912e+01 1.40093024e+01 8.09102638e+05
 4.45210397e+03 1.60780963e+02 3.38771259e+02]

Cluster 1
[1.11787566e+08 2.16933962e+03 1.84829011e+03 3.87903139e+04
 3.36902354e+01 8.62378268e+02 2.15279411e+03]

Cluster 2
[3.17587866e+03 5.04686192e+01 5.07656904e+01 3.94770150e+08
 3.33459463e+04 1.17182814e+04 2.01378904e+04]



### Discusión de resultados – Modelo de aprendizaje no supervisado

Para el experimento de aprendizaje no supervisado se implementó el algoritmo **K-Means**, utilizando las mismas variables numéricas empleadas en el modelo supervisado: *Flow Duration*, *Total Fwd Packet*, *Total Bwd packets*, *Flow Bytes/s*, *Flow Packets/s*, *Packet Length Mean* y *Packet Length Std*. El objetivo fue identificar agrupamientos naturales dentro de los registros de tráfico de red sin utilizar la información contenida en la variable objetivo **Label**.

El modelo fue configurado para generar **tres clusters (k = 3)**. Una vez completado el proceso de entrenamiento, se obtuvo una asignación de grupos para cada registro de la muestra, permitiendo identificar patrones de comportamiento similares entre distintos flujos de red.

La calidad del agrupamiento fue evaluada mediante el coeficiente **Silhouette**, obteniéndose un valor de **0.9899**. Este resultado se considera excelente, ya que valores cercanos a 1 indican que los registros pertenecientes a un mismo cluster presentan una alta similitud interna y, al mismo tiempo, una clara separación respecto a los demás grupos identificados.

Adicionalmente, se analizaron los centros de cada cluster, observándose diferencias importantes en variables como la duración de los flujos, la cantidad de paquetes transmitidos y la tasa de transferencia de datos. Estas diferencias sugieren la existencia de distintos perfiles de comportamiento dentro del tráfico de red analizado. Por ejemplo, algunos grupos presentan flujos de corta duración pero altas tasas de transferencia, mientras que otros muestran conexiones más prolongadas con mayores volúmenes de paquetes intercambiados.

En conjunto, los resultados obtenidos demuestran que el algoritmo K-Means fue capaz de identificar estructuras y patrones relevantes dentro de los datos sin necesidad de utilizar etiquetas previas. Esto confirma la utilidad de las técnicas de aprendizaje no supervisado para explorar grandes conjuntos de datos y descubrir comportamientos que pueden no ser evidentes mediante un análisis tradicional.
